### PR TITLE
refactor!?:  rewrite project runner startup process

### PR DESCRIPTION
**Pull request by barak007**
CI testing

* Separate the coupling between `watch` option to the temp directory isolation
* Remove all sideEffects from the constructor of ProjectRunner (`describe` time) and move them to mocha `before` time
  - Create temp directory
  - Load webpack config
* Better cleanup of temp directory
* Built-in - #2143 (sorry not based on same commit)
* Simplify startup with `prepareTestDirectory` extracted from mocha specific setup
* Use the real path of input directory
* Remove parallel test limit
* disable more cache response options for server 
* improve test speed 
  - no need to wait for network idle
  - one browser server 
 

- [x] should the browser server require be in mocharc ?

### CODE REVIEW

**Comment by idoros** [comment]
This branch still fail on my machine. The only thing that made tests non flaky until now was to reduce the amount of parallel jobs Mocha is running (to 4 in my case).

File: packages/e2e-test-kit/src/project-runner.ts [hunk 1]
```diff
@@ -153,25 +108,28 @@ export class ProjectRunner {
...
-        let done = false;
-        return new Promise<void>((res) => {
-            this.compiler?.hooks.afterDone.tap('waitForRecompile', () => {
-                if (done) {
-                    return;
-                }
-                done = true;
+        return new Promise<void>((res, rej) => {
+            if (!this.compiler) {
+                return rej();
```
Comment (AviVahl) [inline packages/e2e-test-kit/src/project-runner.ts:126]: add new Error and tell the reason

**Comment by AviVahl** [comment]
`throwOnBuildError` used to default to `true`. now it defaults to `false`. Is there anything else less strict now?

File: packages/e2e-test-kit/src/project-runner.ts (hunk 1 above)
Comment (AviVahl) [inline packages/e2e-test-kit/src/project-runner.ts:126]: new Error saying there reason?

**Comment by tzachbon** [comment]
> This branch still fail on my machine. The only thing that made tests non flaky until now was to reduce the amount of parallel jobs Mocha is running (to 4 in my case).

I guess the middle ground here is env variable.

**Comment by idoros** [comment]
> I guess the middle ground here is env variable.

Barak added some improvements since I wrote the comment. Rollup test went from 15s to 5s on my machine, running all the tests went from ~190s down to 70s, at most I get 4-5 chromium processes, and all test are passing without flakiness in the few times I run them. 

Amazing work!

File: packages/e2e-test-kit/src/project-runner.ts [hunk 2]
```diff
@@ -69,109 +52,84 @@ export class ProjectRunner {
...
+        this.prepareTestDirectory();
+        return this.options.watchMode ? this.watch() : this.bundle();
     }
     public async bundle() {
         this.log('Bundle Start');
-        const webpackConfig = this.getWebpackConfig();
+        const webpackConfig = this.loadWebpackConfig();
         const compiler = webpack(webpackConfig);
         this.compiler = compiler;
-        // compiler.run = compiler.run.bind(compiler);
```
Comment (barak007) [inline packages/e2e-test-kit/src/project-runner.ts:122]: remove line

File: packages/e2e-test-kit/src/browser-server.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import { BrowserServer, chromium } from 'playwright-chromium';
+
+let server: BrowserServer;
```
Comment (barak007) [inline packages/e2e-test-kit/src/browser-server.ts:3]: move to this

File: .mocharc.js [hunk 4]
```diff
@@ -1,7 +1,5 @@
-const os = require('os');
-
 module.exports = {
     colors: true,
     'enable-source-maps': true,
-    jobs: Math.min(2, os.cpus().length - 1),
+    require: [require.resolve('@stylable/e2e-test-kit/dist/browser-server.js')],
```
Comment (barak007) [inline .mocharc.js:4]: Two options:
1. Move to package.json to not run in all tests. also add to the webpack-plugin 
2. look at process.cwd and enable when running from root or webpack-plugin

Comment (tzachbon) [reply .mocharc.js:4]: 1

Comment (barak007) [reply .mocharc.js]: I choose 2 because it will give better experience when working on single package.

**Pull request merged**